### ISSUE TITLE
DBSCAN incorrect noise?

### ISSUE THREAD

**Issue by ghost** [annoviko/pyclustering#462]
When I run the code example attached below, one of the points is labelled as noise even though I'm pretty sure it should be part of the cluster. 
![tmp](https://user-images.githubusercontent.com/14364753/45328979-5fbf9f80-b5b2-11e8-9029-8f593690498c.png)

The point in the centre of the circle is a core point (has least 5 points in its neighbourhood), which therefore means that all points within its neighbourhood are part of the cluster (section 3 of http://www.aaai.org/Papers/KDD/1996/KDD96-037.pdf)

I had a look at the python implementation and I believe the issue is that once a point is assigned as a noise point it cannot be changed anymore (since self.__belong[i] in process is set to True, which means that if the point is reached again via a cluster point, it will not be added to the cluster (in __expand_cluster). Changing the position of the point (ix 8) to the end of the data array, kind of confirms this, as the point is then considered as part of the cluster.
![tmp](https://user-images.githubusercontent.com/14364753/45329254-8f22dc00-b5b3-11e8-8fad-2f5935e937a9.png)

```
import numpy as np
import matplotlib.pyplot as plt

from pyclustering.cluster.dbscan import dbscan

def on_picker(event):
    distances = np.abs(np.linalg.norm(data - np.array([event.mouseevent.xdata, event.mouseevent.ydata]), axis=1, ord=2))
    ix = distances.argmin()

    circle = plt.Circle((data[ix, 0], data[ix, 1]), eps, edgecolor='k', linestyle='-', linewidth=0.5, fill=None)

    ax.add_artist(circle)
    event.canvas.draw()

data = np.array([[3.84883118, 7.98287654],
                 [6.15957575, 0.93742612],
                 [0.58865177, 7.34201309],
                 [6.95169446, 6.34775643],
                 [3.64920659, 1.43836441],
                 [4.98745463, 1.48505],
                 [1.50897943, 0.66372945],
                 [1.7165765, 6.51867617],
                 # [3.63916014, 4.03335428],        # Point in question
                 [1.64755443, 6.91168503],
                 [5.18619498, 5.01467501],
                 [5.22841242, 4.66132755],
                 [4.46416522, 4.99414511],
                 [5.74935967, 5.19903379],
                 [4.71599415, 4.88615725],
                 [5.24744658, 4.56903919],
                 [5.19049452, 5.41090111],
                 [4.46223735, 4.43709619],
                 [4.06165299, 5.64246575],
                 [4.13648305, 5.0203021],
                 [5.0931739, 5.90484805],
                 [5.27248291, 5.10978874],
                 [4.98665978, 5.02617475],
                 [4.49543272, 3.6342554],
                 [5.49451608, 4.56799783],
                 [5.26142474, 4.99616694],
                 [5.5543285, 5.59529684],
                 [4.15108339, 4.31406758],
                 [5.12194741, 4.93568024],
                 [4.77956745, 4.73403704],
                 [5.58123982, 5.5657367],
                 [5.06730322, 4.57878869],
                 [4.58515378, 5.18852395],
                 [5.49669378, 4.8901156],
                 [5.42696559, 4.71035229],
                 [3.5934053, 4.76841482],
                 [5.60826032, 5.61143477],
                 [6.23623543, 4.93332256],
                 [5.25518697, 4.35253186],
                 [3.63467815, 5.25080737],
                 # [3.63467815, 5.25080737]])
                 [3.63916014, 4.03335428]])

eps, minPts = 1.0, 5
dbscan_i = dbscan(data, eps, minPts, False)
dbscan_i.process()

labels = np.zeros(data.shape[0], dtype=np.int32)
for ix, cluster_indices in enumerate(dbscan_i.get_clusters()):
    labels[cluster_indices] = ix + 1

fig = plt.figure()
ax = fig.add_subplot(111)
ax.axis("equal")
for label in labels:
    mask = labels == label
    if label == 0:
        ax.scatter(data[mask, 0], data[mask, 1], s=10, c='k', picker=True)
    else:
        ax.scatter(data[mask, 0], data[mask, 1], s=10, picker=True)
ax.set_title("Eps {}, MinPts {}".format(eps, minPts))

fig.canvas.mpl_connect("pick_event", on_picker)

plt.show()
```

Happy to create a PR that fixes this (at least for the python implementation), if wanted?

Thanks

**Comment by annoviko** [comment]
Hello, @Clus10,

Yes, you are correct, self.__belong[i] never returns to False, only True value can be set. I have uncommented the issue point and run your code example, and  found out that point is not marked as a noise, it belongs to cluster (I have run it in two configuration: using C++ implementation when CCORE flag is True, and pure Python implementation when CCORE flag is False):
![1](https://user-images.githubusercontent.com/6785869/45348223-bc24be00-b5b6-11e8-9b66-6966bdc5450e.png)
![2](https://user-images.githubusercontent.com/6785869/45348226-be871800-b5b6-11e8-8fbf-e54e7cdf5fb8.png)

**Comment by Clus10** [comment]
Hey @annoviko, sorry I wasn't quite clear enough what I meant with the data.
So if you run the example with this dataset
```
data = np.array([[3.84883118, 7.98287654],
                 [6.15957575, 0.93742612],
                 [0.58865177, 7.34201309],
                 [6.95169446, 6.34775643],
                 [3.64920659, 1.43836441],
                 [4.98745463, 1.48505],
                 [1.50897943, 0.66372945],
                 [1.7165765, 6.51867617],
                 [3.63916014, 4.03335428],        # Point in question
                 [1.64755443, 6.91168503],
                 [5.18619498, 5.01467501],
                 [5.22841242, 4.66132755],
                 [4.46416522, 4.99414511],
                 [5.74935967, 5.19903379],
                 [4.71599415, 4.88615725],
                 [5.24744658, 4.56903919],
                 [5.19049452, 5.41090111],
                 [4.46223735, 4.4370961],
                 [4.06165299, 5.64246575],
                 [4.13648305, 5.0203021],
                 [5.0931739, 5.90484805],
                 [5.27248291, 5.10978874],
                 [4.98665978, 5.02617475],
                 [4.49543272, 3.6342554],
                 [5.49451608, 4.56799783],
                 [5.26142474, 4.99616694],
                 [5.5543285, 5.59529684],
                 [4.15108339, 4.31406758],
                 [5.12194741, 4.93568024],
                 [4.77956745, 4.73403704],
                 [5.58123982, 5.5657367],
                 [5.06730322, 4.57878869],
                 [4.58515378, 5.18852395],
                 [5.49669378, 4.8901156],
                 [5.42696559, 4.71035229],
                 [3.5934053, 4.76841482],
                 [5.60826032, 5.61143477],
                 [6.23623543, 4.93332256],
                 [5.25518697, 4.35253186],
                 [3.63467815, 5.25080737],
                 [3.63467815, 5.25080737]])
```
the point will not be part of the cluster. However if you use this one

```
data = np.array([[3.84883118, 7.98287654],
                 [6.15957575, 0.93742612],
                 [0.58865177, 7.34201309],
                 [6.95169446, 6.34775643],
                 [3.64920659, 1.43836441],
                 [4.98745463, 1.48505],
                 [1.50897943, 0.66372945],
                 [1.7165765, 6.51867617],
                 [1.64755443, 6.91168503],
                 [5.18619498, 5.01467501],
                 [5.22841242, 4.66132755],
                 [4.46416522, 4.99414511],
                 [5.74935967, 5.19903379],
                 [4.71599415, 4.88615725],
                 [5.24744658, 4.56903919],
                 [5.19049452, 5.41090111],
                 [4.46223735, 4.4370961],
                 [4.06165299, 5.64246575],
                 [4.13648305, 5.0203021],
                 [5.0931739, 5.90484805],
                 [5.27248291, 5.10978874],
                 [4.98665978, 5.02617475],
                 [4.49543272, 3.6342554],
                 [5.49451608, 4.56799783],
                 [5.26142474, 4.99616694],
                 [5.5543285, 5.59529684],
                 [4.15108339, 4.31406758],
                 [5.12194741, 4.93568024],
                 [4.77956745, 4.73403704],
                 [5.58123982, 5.5657367],
                 [5.06730322, 4.57878869],
                 [4.58515378, 5.18852395],
                 [5.49669378, 4.8901156],
                 [5.42696559, 4.71035229],
                 [3.5934053, 4.76841482],
                 [5.60826032, 5.61143477],
                 [6.23623543, 4.93332256],
                 [5.25518697, 4.35253186],
                 [3.63467815, 5.25080737],
                 [3.63467815, 5.25080737],
                 [3.63916014, 4.03335428]])  # Point in question
```

the point will be part of the cluster. The only difference between the two datasets is that I moved the position of the point in the array.

I use python 3.6.5 and pyclustering (0.8.1)

**Comment by Clus10** [comment]
Hey @annoviko, where you able to reproduce this?

**Comment by annoviko** [comment]
Hello, @Clus10 ,
Thank you for your description, I have reproduced the problem. I will try to find out what is wrong and correct it.

**Comment by annoviko** [comment]
@Clus10 , I have sort this problem out for python and C++ parts, thank you for your detailed report again, it was really useful and helps to find out root cause of the problem. Correction is available on the branch 0.8.dev. You need to rebuild C++ part of the library for linux or windows to use corrected version of the core. Here is an instruction how to build core of the library: https://github.com/annoviko/pyclustering/wiki/Core-of-the-PyClustering.
